Description: Fix linkage for GNU/Hurd and GNU/kFreeBSD
Author: Jonas Smedegaard <dr@jones.dk>
Last-Update: 2013-09-26

--- a/uwsgiconfig.py
+++ b/uwsgiconfig.py
@@ -598,8 +598,10 @@
 
         self.ldflags = os.environ.get("LDFLAGS", "").split()
         self.libs = ['-lpthread', '-lm', '-rdynamic']
-        if uwsgi_os == 'Linux':
+        if uwsgi_os in ('Linux', 'GNU', 'GNU/kFreeBSD'):
             self.libs.append('-ldl')
+        if uwsgi_os == 'GNU/kFreeBSD':
+            self.libs.append('-lbsd')
 
         # check for inherit option
         inherit = self.get('inherit')
@@ -953,7 +955,7 @@
         if self.has_include('uuid/uuid.h'):
             has_uuid = True
             self.cflags.append("-DUWSGI_UUID")
-            if uwsgi_os == 'Linux' or uwsgi_os.startswith('CYGWIN') or os.path.exists('/usr/lib/libuuid.so') or os.path.exists('/usr/local/lib/libuuid.so') or os.path.exists('/usr/lib64/libuuid.so') or os.path.exists('/usr/local/lib64/libuuid.so'):
+            if uwsgi_os in ('Linux', 'GNU', 'GNU/kFreeBSD') or uwsgi_os.startswith('CYGWIN') or os.path.exists('/usr/lib/libuuid.so') or os.path.exists('/usr/local/lib/libuuid.so') or os.path.exists('/usr/lib64/libuuid.so') or os.path.exists('/usr/local/lib64/libuuid.so'):
                 self.libs.append('-luuid')
 
         if self.get('append_version'):
